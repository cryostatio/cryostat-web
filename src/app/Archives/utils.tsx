/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { RecordingDirectory, Target } from '@app/Shared/Services/api.types';

export const includesDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): boolean => {
  return arr.some((t) => t.connectUrl === dir.connectUrl);
};

export const indexOfDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): number => {
  let index = -1;
  arr.forEach((d, idx) => {
    if (d.connectUrl === dir.connectUrl) {
      index = idx;
    }
  });
  return index;
};

export const getTargetFromDirectory = (dir: RecordingDirectory): Target => {
  return {
    connectUrl: dir.connectUrl,
    alias: dir.jvmId,
    labels: [],
    annotations: {
      cryostat: [],
      platform: [],
    },
  };
};
