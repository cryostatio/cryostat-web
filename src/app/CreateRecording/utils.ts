/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { TemplateType } from '@app/Shared/Services/api.types';
import { EventTemplateIdentifier } from './types';

export const RecordingNamePattern = /^[\w_]+$/;

export const DurationPattern = /^[1-9][0-9]*$/;

export const isRecordingNameValid = (name: string) => RecordingNamePattern.test(name);

export const isDurationValid = (duration: number) => DurationPattern.test(`${duration}`);

export const templateFromEventSpecifier = (eventSpecifier: string): EventTemplateIdentifier | undefined => {
  const regex = /^template=([a-zA-Z0-9]+)(?:,type=([a-zA-Z0-9]+))?$/im;
  if (eventSpecifier && regex.test(eventSpecifier)) {
    const parts = regex.exec(eventSpecifier);
    if (parts) {
      return {
        name: parts[1],
        type: parts[2] as TemplateType,
      };
    }
  }
  return undefined;
};
