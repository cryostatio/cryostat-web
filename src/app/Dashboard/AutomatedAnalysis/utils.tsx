/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { AnalysisResult, Evaluation } from '@app/Shared/Services/api.types';
import { Stack, StackItem, Label, Title, Text } from '@patternfly/react-core';
import _ from 'lodash';

export const transformAADescription = (result: AnalysisResult): JSX.Element => {
  const format = (s: Evaluation): JSX.Element => {
    if (typeof s === 'string') {
      return <Text>{s}</Text>;
    }
    if (Array.isArray(s)) {
      return (
        <Stack>
          {s.map((e) => (
            <StackItem key={e.setting}>
              <Title headingLevel={'h6'}>{e.setting}</Title>
              <Label>
                {e.name}={e.value}
              </Label>
            </StackItem>
          ))}
        </Stack>
      );
    }
    throw `Unrecognized item: ${s}`;
  };
  return (
    <div>
      {Object.entries(result.evaluation || {}).map(([k, v]) =>
        v && v.length ? (
          <div key={k}>
            <span>
              <Title headingLevel={'h5'}>{_.capitalize(k)}</Title>
              {format(result.evaluation[k])}
            </span>
            <br />
          </div>
        ) : (
          <div key={k}></div>
        ),
      )}
    </div>
  );
};
