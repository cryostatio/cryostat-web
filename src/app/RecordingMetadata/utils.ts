/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { KeyValue } from '@app/Shared/Services/api.types';
import { ValidatedOptions } from '@patternfly/react-core';
import { Observable, from } from 'rxjs';

export const isEqualLabel = (a: KeyValue, b: KeyValue) => {
  return a.key === b.key && a.value === b.value;
};

export const includesLabel = (arr: KeyValue[], searchLabel: KeyValue) => {
  return arr.some((l) => isEqualLabel(searchLabel, l));
};

export const parseLabelsFromFile = (file: File): Observable<KeyValue[]> => {
  return from(
    file
      .text()
      .then(JSON.parse)
      .then((obj) => {
        const labels: KeyValue[] = [];
        const labelObj = obj['labels'];
        if (labelObj) {
          Object.keys(labelObj).forEach((key) => {
            labels.push({
              key: key,
              value: labelObj[key],
            });
          });
          return labels;
        }
        throw new Error('No labels found in file');
      }),
  );
};

export const getLabelDisplay = (label: KeyValue) => `${label.key}:${label.value}`;

export const LabelPattern = /^\S+$/;

export const getValidatedOption = (isValid: boolean) => {
  return isValid ? ValidatedOptions.success : ValidatedOptions.error;
};

export const matchesLabelSyntax = (l: KeyValue) => {
  return l && LabelPattern.test(l.key) && LabelPattern.test(l.value);
};
