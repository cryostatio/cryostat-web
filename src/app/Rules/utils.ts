/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ruleObjKeys } from './types';

export const isRule = (obj: object): boolean => {
  for (const key of ruleObjKeys) {
    if (!Object.prototype.hasOwnProperty.call(obj, key)) {
      return false;
    }
  } // Ignore unknown fields
  return true;
};

// FIXME check if this is correct/matches backend name validation
export const RuleNamePattern = /^[\w_]+$/;

export const isRuleNameValid = (name: string) => RuleNamePattern.test(name);

export const CEL_SPEC_HREF = 'https://github.com/google/cel-spec/blob/master/doc/langdef.md';
