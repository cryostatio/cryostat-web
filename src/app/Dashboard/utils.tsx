/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import cryostatLogo from '@app/assets/cryostat_icon_rgb_default.svg';
import cryostatLogoDark from '@app/assets/cryostat_icon_rgb_reverse.svg';
import { dashboardConfigDeleteCardIntent } from '@app/Shared/Redux/ReduxStore';
import { FeatureLevel } from '@app/Shared/Services/service.types';
import { withThemedIcon } from '@app/utils/withThemedIcon';
import { useCryostatTranslation } from '@i18n/i18nextUtil';
import { Button, ButtonVariant } from '@patternfly/react-core';
import { FileIcon, UnknownIcon, UserIcon } from '@patternfly/react-icons';
import { nanoid } from '@reduxjs/toolkit';
import { TFunction } from 'i18next';
import * as React from 'react';
import { useDispatch } from 'react-redux';
import { AutomatedAnalysisCardDescriptor } from './AutomatedAnalysis/AutomatedAnalysisCard';
import { JFRMetricsChartCardDescriptor } from './Charts/jfr/JFRMetricsChartCard';
import { MBeanMetricsChartCardDescriptor } from './Charts/mbean/MBeanMetricsChartCard';
import { DiagnosticsCardDescriptor } from './Diagnostics/DiagnosticsCard';
import { JvmDetailsCardDescriptor } from './JvmDetails/JvmDetailsCard';
import {
  SerialLayoutTemplate,
  SerialCardConfig,
  LayoutTemplateVersion,
  LayoutTemplateController,
  LayoutTemplateVendor,
  DashboardLayout,
  LayoutTemplate,
  LayoutTemplateRecord,
  CardConfig,
  CardValidationResult,
  ValidationError,
  DashboardCardDescriptor,
} from './types';

export const DashboardLayoutNamePattern = /^[a-zA-Z0-9_.-]+( [a-zA-Z0-9_.-]+)*$/;
export const LayoutTemplateDescriptionPattern = /^[a-zA-Z0-9\s.,\-'";?!@#$%^&*()[\]_+=:{}]*$/;

export const mockSerialLayoutTemplate: SerialLayoutTemplate = {
  name: 'Default',
  description: 'Default.',
  cards: [] as SerialCardConfig[],
  version: LayoutTemplateVersion['v2.3'],
};

// use a provider
export const LayoutTemplateContext = React.createContext<LayoutTemplateController>({
  selectedTemplate: undefined,
  setSelectedTemplate: () => undefined,
  isUploadModalOpen: false,
  setIsUploadModalOpen: () => undefined,
});

export const iconify = (vendor: LayoutTemplateVendor): React.ReactNode => {
  switch (vendor) {
    case LayoutTemplateVendor.CRYOSTAT:
      return <ThemedCryostatLogo />;
    case LayoutTemplateVendor.BLANK:
      return <FileIcon style={{ paddingRight: '0.3rem' }} />;
    case LayoutTemplateVendor.USER:
      return <UserIcon />;
    default:
      return <UnknownIcon />;
  }
};

export const ThemedCryostatLogo = withThemedIcon(cryostatLogo, cryostatLogoDark, 'Cryostat Logo');

export const templatize = (layout: DashboardLayout, name: string, desc?: string): LayoutTemplate => {
  return {
    name: name,
    description: desc || `Custom layout template.`,
    cards: layout.cards.map((card) => {
      const { id: _id, ...cardWithoutId } = card;
      return { ...cardWithoutId };
    }) as SerialCardConfig[],
    vendor: LayoutTemplateVendor.USER,
    version: LayoutTemplateVersion['v2.3'],
  } as LayoutTemplate;
};

export const layoutize = (template: LayoutTemplate, name: string): DashboardLayout => {
  return {
    name: name,
    cards: template.cards.map((card) => {
      return {
        ...card,
        id: `${card.name}-${nanoid()}`,
      };
    }),
    favorite: false,
  } as DashboardLayout;
};

export const recordToLayoutTemplate = (
  record: LayoutTemplateRecord,
  allTemplates: LayoutTemplate[],
): LayoutTemplate | undefined => {
  return allTemplates.find((template) => template.name === record.name && template.vendor === record.vendor);
};

export const getUniqueIncrementingName = (init = 'Custom', names: string[]): string => {
  let name: string;
  let i = 1;
  do {
    name = `${init}${i}`;
    i++;
  } while (names.includes(name));
  return name;
};

export const hasCardDescriptorByName = (name: string): boolean => {
  for (const choice of getDashboardCards()) {
    if (choice.component.cardComponentName === name) {
      return true;
    }
  }
  return false;
};

export const getCardDescriptorByName = (name: string): DashboardCardDescriptor => {
  for (const choice of getDashboardCards()) {
    if (choice.component.cardComponentName === name) {
      return choice;
    }
  }
  throw new Error(`Unknown card type selection: ${name}`);
};

export const hasCardDescriptorByTitle = (title: string, t: TFunction): boolean => {
  for (const choice of getDashboardCards()) {
    if (t(choice.title) === title) {
      return true;
    }
  }
  return false;
};

export const getCardDescriptorByTitle = (title: string, t: TFunction): DashboardCardDescriptor => {
  for (const choice of getDashboardCards()) {
    if (t(choice.title) === title) {
      return choice;
    }
  }
  throw new Error(`Unknown card type selection: ${title}`);
};

export const getDashboardCards: (featureLevel?: FeatureLevel) => DashboardCardDescriptor[] = (
  featureLevel = FeatureLevel.DEVELOPMENT,
) => {
  const cards = [
    JvmDetailsCardDescriptor,
    AutomatedAnalysisCardDescriptor,
    JFRMetricsChartCardDescriptor,
    MBeanMetricsChartCardDescriptor,
    DiagnosticsCardDescriptor,
  ];
  return cards.filter((card) => card.featureLevel >= featureLevel);
};

export const RemoveCardAction: React.FC<{ cardIndex: number }> = ({ cardIndex }) => {
  const dispatch = useDispatch();
  const { t } = useCryostatTranslation();

  const handleClick = React.useCallback(() => {
    dispatch(dashboardConfigDeleteCardIntent(cardIndex));
  }, [dispatch, cardIndex]);

  return (
    <Button onClick={handleClick} variant={ButtonVariant.danger}>
      {t('REMOVE')}
    </Button>
  );
};

export const validateCardConfig = ({ name, props }: CardConfig, cardIndex: number): CardValidationResult => {
  // Unsupported card type
  if (!hasCardDescriptorByName(name)) {
    return {
      errors: [
        {
          message: (
            <>
              Unknown card type: <code>{name}</code>.
            </>
          ),
        },
      ],
      callForAction: <RemoveCardAction cardIndex={cardIndex} />,
    };
  }

  const errs: ValidationError[] = [];

  const { propControls } = getCardDescriptorByName(name);
  const configPropKeys = Object.entries(props);

  // Missing props
  propControls
    .map((ctrl) => ctrl.key)
    .forEach((propKey) => {
      const matched = configPropKeys.find(([key, _]) => key === propKey);
      if (!matched) {
        errs.push({
          message: (
            <>
              Missing card property: <code>{propKey}</code>.
            </>
          ),
        });
      }
    });

  // Invalid prop valies
  configPropKeys.forEach(([propKey, propValue]) => {
    const matched = propControls.find((crtl) => crtl.key === propKey);
    if (matched) {
      const { values, extras, kind } = matched;
      // FIXME: Check dynamic values
      let err: ValidationError | undefined;
      if (propValue === undefined || propValue === null) {
        err = {
          message: (
            <>
              Undefined value for card property: <code>{propKey}</code>
            </>
          ),
        };
      } else if (values && Array.isArray(values) && !values.some((v) => v === propValue)) {
        err = {
          message: (
            <>
              Invalid value <code>{propValue}</code> for card property: <code>{propKey}</code>.
            </>
          ),
        };
      } else if (extras) {
        if (kind === 'number') {
          const valAsNum = Number(propValue);
          if (isNaN(valAsNum)) {
            err = {
              message: (
                <>
                  Numeric value expected but <code>{typeof propValue}</code> was given for: <>{propKey}</>.
                </>
              ),
            };
          } else {
            if (
              (extras.min !== undefined && valAsNum < extras.min) ||
              (extras.max !== undefined && valAsNum > extras.max)
            ) {
              err = {
                message: (
                  <>
                    Value exceeding limit for card property: <code>{propKey}</code>.
                  </>
                ),
              };
            }
          }
        }
      }
      err && errs.push(err);
    }
    // Unknown props are ignored
  });

  return {
    errors: errs,
    callForAction: errs.length ? <RemoveCardAction cardIndex={cardIndex} /> : undefined,
  };
};
