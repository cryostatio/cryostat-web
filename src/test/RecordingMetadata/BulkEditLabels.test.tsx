/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { BulkEditLabels } from '@app/RecordingMetadata/BulkEditLabels';
import {
  ArchivedRecording,
  ActiveRecording,
  RecordingState,
  NotificationMessage,
  Target,
} from '@app/Shared/Services/api.types';
import { defaultServices } from '@app/Shared/Services/Services';
import '@testing-library/jest-dom';
import { cleanup, screen } from '@testing-library/react';
import { of } from 'rxjs';
import { renderSnapshot, render } from '../utils';

jest.mock('@patternfly/react-core', () => ({
  ...jest.requireActual('@patternfly/react-core'),
  Tooltip: ({ t }) => <>{t}</>,
}));

const mockConnectUrl = 'service:jmx:rmi://someUrl';
const mockJvmId = 'id';
const mockTarget: Target = {
  connectUrl: mockConnectUrl,
  alias: 'fooTarget',
  jvmId: mockJvmId,
  labels: [],
  annotations: { cryostat: [], platform: [] },
};

const mockRecordingLabels = [
  {
    key: 'someLabel',
    value: 'someValue',
  },
];

const mockArchivedRecording: ArchivedRecording = {
  name: 'someArchivedRecording_some_random',
  downloadUrl: 'http://downloadUrl',
  reportUrl: 'http://reportUrl',
  metadata: { labels: mockRecordingLabels },
  size: 2048,
  archivedTime: 2048,
};

const mockActiveRecording: ActiveRecording = {
  name: 'someActiveRecording',
  downloadUrl: 'http://downloadUrl',
  reportUrl: 'http://reportUrl',
  metadata: { labels: mockRecordingLabels },
  startTime: 1234567890,
  id: 0,
  state: RecordingState.RUNNING,
  duration: 0,
  continuous: false,
  toDisk: false,
  maxSize: 0,
  maxAge: 0,
};

const mockActiveLabelsNotification = {
  message: {
    target: mockConnectUrl,
    recordingName: 'someActiveRecording',
    jvmId: mockJvmId,
    metadata: { labels: { someLabel: 'someValue', someNewLabel: 'someNewValue' } },
  },
} as NotificationMessage;

const mockActiveRecordingResponse = [mockActiveRecording];

const mockArchivedLabelsNotification = {
  message: {
    target: mockConnectUrl,
    recordingName: 'someArchivedRecording_some_random',
    jvmId: mockJvmId,
    metadata: { labels: { someLabel: 'someValue', someNewLabel: 'someNewValue' } },
  },
} as NotificationMessage;

const mockArchivedRecordingsResponse = {
  data: {
    targetNodes: [
      {
        recordings: {
          archived: {
            data: [mockArchivedRecording] as ArchivedRecording[],
          },
        },
      },
    ],
  },
};

jest.spyOn(defaultServices.target, 'target').mockReturnValue(of(mockTarget));
jest.spyOn(defaultServices.api, 'graphql').mockReturnValue(of(mockArchivedRecordingsResponse));
jest.spyOn(defaultServices.api, 'doGet').mockReturnValue(of(mockActiveRecordingResponse));
jest
  .spyOn(defaultServices.notificationChannel, 'messages')
  .mockReturnValueOnce(of()) // renders correctly
  .mockReturnValueOnce(of()) // should display read-only labels from selected recordings
  .mockReturnValueOnce(of()) // should display editable labels form when in edit mode
  .mockReturnValueOnce(of()) // should not display labels for unchecked recordings
  .mockReturnValueOnce(of(mockActiveLabelsNotification)) // should update the target recording labels after receiving a notification
  .mockReturnValueOnce(of(mockArchivedLabelsNotification)) // should update the archived recording labels after receiving a notification
  .mockReturnValue(of());

describe('<BulkEditLabels />', () => {
  let activeCheckedIndices: number[];
  let archivedCheckedIndices: number[];
  let emptycheckIndices: number[];

  beforeEach(() => {
    activeCheckedIndices = [mockActiveRecording.id];
    archivedCheckedIndices = [-553224758]; // Hash code of "someArchivedRecording_some_random"
    emptycheckIndices = [];
  });

  afterEach(cleanup);

  it('renders correctly', async () => {
    const tree = await renderSnapshot({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <BulkEditLabels checkedIndices={activeCheckedIndices} isTargetRecording={true} />,
          },
        ],
      },
    });
    expect(tree?.toJSON()).toMatchSnapshot();
  });

  it('should display read-only labels from selected recordings', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <BulkEditLabels checkedIndices={activeCheckedIndices} isTargetRecording={true} />,
          },
        ],
      },
    });

    const label = screen.getByLabelText('someLabel: someValue');
    expect(label).toBeInTheDocument();
    expect(label).toBeVisible();
    expect(label.onclick).toBeNull();

    const addLabelButton = screen.queryByRole('button', { name: 'Add Label' });
    expect(addLabelButton).not.toBeInTheDocument();

    const editButton = screen.getByRole('button', { name: 'Edit Labels' });
    expect(editButton).toBeInTheDocument();
    expect(editButton).toBeVisible();
    expect(editButton).not.toBeDisabled();
  });

  it('should not display labels for unchecked recordings', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <BulkEditLabels checkedIndices={emptycheckIndices} isTargetRecording={true} />,
          },
        ],
      },
    });

    expect(screen.queryByText('someLabel')).not.toBeInTheDocument();
    expect(screen.queryByText('someValue')).not.toBeInTheDocument();

    const placeHolder = screen.getByText('-');
    expect(placeHolder).toBeInTheDocument();
    expect(placeHolder).toBeVisible();
  });

  it('should display editable labels form when in edit mode', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <BulkEditLabels checkedIndices={activeCheckedIndices} isTargetRecording={true} />,
          },
        ],
      },
    });

    const editButton = screen.getByRole('button', { name: 'Edit Labels' });
    expect(editButton).toBeInTheDocument();
    expect(editButton).toBeVisible();

    await user.click(editButton);

    const addLabelButton = screen.getByRole('button', { name: 'Add Label' });
    expect(addLabelButton).toBeInTheDocument();
    expect(addLabelButton).toBeVisible();

    const labelKeyInput = screen.getAllByDisplayValue('someLabel')[0];
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toBeInTheDocument();
    expect(labelKeyInput).toBeVisible();

    const labelValueInput = screen.getAllByDisplayValue('someValue')[0];
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toBeInTheDocument();
    expect(labelValueInput).toBeVisible();

    const saveButton = screen.getByText('Save');
    expect(saveButton).toBeInTheDocument();
    expect(saveButton).toBeVisible();

    const cancelButton = screen.getByText('Cancel');
    expect(cancelButton).toBeInTheDocument();
    expect(cancelButton).toBeVisible();
  });

  it('should update the target recording labels after receiving a notification', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <BulkEditLabels checkedIndices={activeCheckedIndices} isTargetRecording={true} />,
          },
        ],
      },
    });

    const newLabel = screen.getByLabelText('someNewLabel: someNewValue');
    expect(newLabel).toBeInTheDocument();
    expect(newLabel).toBeVisible();

    const oldLabel = screen.getByLabelText('someLabel: someValue');
    expect(oldLabel).toBeInTheDocument();
    expect(oldLabel).toBeVisible();
  });

  it('should update the archived recording labels after receiving a notification', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <BulkEditLabels checkedIndices={archivedCheckedIndices} isTargetRecording={false} />,
          },
        ],
      },
    });

    const newLabel = screen.getByLabelText('someNewLabel: someNewValue');
    expect(newLabel).toBeInTheDocument();
    expect(newLabel).toBeVisible();

    const oldLabel = screen.getByLabelText('someLabel: someValue');
    expect(oldLabel).toBeInTheDocument();
    expect(oldLabel).toBeVisible();
  });

  it('should return to read-only view when edited labels are cancelled', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <BulkEditLabels checkedIndices={archivedCheckedIndices} isTargetRecording={false} />,
          },
        ],
      },
    });

    let editButton = screen.getByRole('button', { name: 'Edit Labels' });
    expect(editButton).toBeInTheDocument();
    expect(editButton).toBeVisible();

    await user.click(editButton);

    const addLabelButton = screen.getByRole('button', { name: 'Add Label' });
    expect(addLabelButton).toBeInTheDocument();
    expect(addLabelButton).toBeVisible();

    const labelKeyInput = screen.getAllByDisplayValue('someLabel')[0];
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toBeInTheDocument();
    expect(labelKeyInput).toBeVisible();

    const labelValueInput = screen.getAllByDisplayValue('someValue')[0];
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toBeInTheDocument();
    expect(labelValueInput).toBeVisible();

    const saveButton = screen.getByText('Save');
    expect(saveButton).toBeInTheDocument();
    expect(saveButton).toBeVisible();

    const cancelButton = screen.getByText('Cancel');
    expect(cancelButton).toBeInTheDocument();
    expect(cancelButton).toBeVisible();

    await user.click(cancelButton);

    editButton = screen.getByRole('button', { name: 'Edit Labels' });
    expect(editButton).toBeInTheDocument();
    expect(editButton).toBeVisible();
    expect(addLabelButton).not.toBeInTheDocument();
  });

  it('should save target recording labels when Save is clicked', async () => {
    const saveRequestSpy = jest
      .spyOn(defaultServices.api, 'postTargetRecordingMetadata')
      .mockReturnValue(of([mockActiveRecording]));
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <BulkEditLabels checkedIndices={activeCheckedIndices} isTargetRecording={true} />,
          },
        ],
      },
    });

    const editButton = screen.getByRole('button', { name: 'Edit Labels' });
    expect(editButton).toBeInTheDocument();
    expect(editButton).toBeVisible();

    await user.click(editButton);

    const addLabelButton = screen.getByRole('button', { name: 'Add Label' });
    expect(addLabelButton).toBeInTheDocument();
    expect(addLabelButton).toBeVisible();

    const labelKeyInput = screen.getAllByDisplayValue('someLabel')[0];
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toBeInTheDocument();
    expect(labelKeyInput).toBeVisible();

    const labelValueInput = screen.getAllByDisplayValue('someValue')[0];
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toBeInTheDocument();
    expect(labelValueInput).toBeVisible();

    const saveButton = screen.getByText('Save');
    expect(saveButton).toBeInTheDocument();
    expect(saveButton).toBeVisible();

    const cancelButton = screen.getByText('Cancel');
    expect(cancelButton).toBeInTheDocument();
    expect(cancelButton).toBeVisible();

    await user.click(saveButton);

    expect(saveRequestSpy).toHaveBeenCalledTimes(1);
  });

  it('should save archived recording labels when Save is clicked', async () => {
    const saveRequestSpy = jest
      .spyOn(defaultServices.api, 'postRecordingMetadata')
      .mockReturnValue(of([mockArchivedRecording]));
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <BulkEditLabels checkedIndices={archivedCheckedIndices} isTargetRecording={false} />,
          },
        ],
      },
    });

    const editButton = screen.getByRole('button', { name: 'Edit Labels' });
    expect(editButton).toBeInTheDocument();
    expect(editButton).toBeVisible();

    await user.click(editButton);

    const addLabelButton = screen.getByRole('button', { name: 'Add Label' });
    expect(addLabelButton).toBeInTheDocument();
    expect(addLabelButton).toBeVisible();

    const labelKeyInput = screen.getAllByDisplayValue('someLabel')[0];
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toHaveClass('pf-c-form-control');
    expect(labelKeyInput).toBeInTheDocument();
    expect(labelKeyInput).toBeVisible();

    const labelValueInput = screen.getAllByDisplayValue('someValue')[0];
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toHaveClass('pf-c-form-control');
    expect(labelValueInput).toBeInTheDocument();
    expect(labelValueInput).toBeVisible();

    const saveButton = screen.getByText('Save');
    expect(saveButton).toBeInTheDocument();
    expect(saveButton).toBeVisible();

    const cancelButton = screen.getByText('Cancel');
    expect(cancelButton).toBeInTheDocument();
    expect(cancelButton).toBeVisible();

    await user.click(saveButton);

    expect(saveRequestSpy).toHaveBeenCalledTimes(1);
  });
});
