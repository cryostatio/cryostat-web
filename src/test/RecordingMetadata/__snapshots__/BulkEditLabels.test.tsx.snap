// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BulkEditLabels /> renders correctly 1`] = `
<div
  className="pf-v5-l-stack pf-m-gutter"
>
  <div
    className="pf-v5-l-stack__item"
  >
    <div
      className="pf-v5-l-split pf-m-gutter"
    >
      <div
        className="pf-v5-l-split__item"
      >
        <p
          className=""
          data-ouia-component-id="OUIA-Generated-Text-1"
          data-ouia-component-type="PF5/Text"
          data-ouia-safe={true}
          data-pf-content={true}
        >
          Edit Recording Labels
        </p>
      </div>
      <div
        className="pf-v5-l-split__item"
      />
    </div>
  </div>
  <div
    className="pf-v5-l-stack__item"
  >
    <span
      aria-label="someLabel: someValue"
      className="pf-v5-c-label"
    >
      <span
        className="pf-v5-c-label__content"
      >
        <span
          className="pf-v5-c-label__text"
        >
          someLabel: someValue
        </span>
      </span>
    </span>
  </div>
  <div
    className="pf-v5-l-stack__item"
  >
    <button
      aria-disabled={false}
      aria-label="Edit Labels"
      className="pf-v5-c-button pf-m-secondary"
      data-ouia-component-id="OUIA-Generated-Button-secondary-1"
      data-ouia-component-type="PF5/Button"
      data-ouia-safe={true}
      disabled={false}
      onClick={[Function]}
      role={null}
      type="button"
    >
      Edit
    </button>
  </div>
</div>
`;
